azure/services/natgateways/natgateways.go

```diff
@@ -0,0 +1,148 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package natgateways
+
+import (
+	"context"
+
+	"github.com/Azure/azure-sdk-for-go/services/network/mgmt/2019-06-01/network"
+	"github.com/Azure/go-autorest/autorest/to"
+	"github.com/go-logr/logr"
+	"github.com/pkg/errors"
+
+	infrav1 "sigs.k8s.io/cluster-api-provider-azure/api/v1alpha4"
+	"sigs.k8s.io/cluster-api-provider-azure/azure"
+	"sigs.k8s.io/cluster-api-provider-azure/util/tele"
+)
+
+// NatGatewayScope defines the scope interface for nat gateway service.
+type NatGatewayScope interface {
+	logr.Logger
+	azure.ClusterScoper
+	NatGatewaySpecs() []azure.NatGatewaySpec
+	SetNodeNatGateway(natGateway infrav1.NatGateway)
+}
+
+// Service provides operations on azure resources.
+type Service struct {
+	Scope NatGatewayScope
+	client
+}
+
+// New creates a new service.
+func New(scope NatGatewayScope) *Service {
+	return &Service{
+		Scope:  scope,
+		client: newClient(scope),
+	}
+}
+
+// Reconcile gets/creates/updates a nat gateway.
+// Only when the Nat Gateway 'Name' property is defined we create the Nat Gateway: it's opt-in.
+func (s *Service) Reconcile(ctx context.Context) error {
+	ctx, span := tele.Tracer().Start(ctx, "natgateways.Service.Reconcile")
+	defer span.End()
+
+	if !s.Scope.Vnet().IsManaged(s.Scope.ClusterName()) {
+		s.Scope.V(4).Info("Skipping nat gateways reconcile in custom vnet mode")
+		return nil
+	}
+
+	for _, natGatewaySpec := range s.Scope.NatGatewaySpecs() {
+		existingNatGateway, err := s.getExisting(ctx, natGatewaySpec)
+
+		switch {
+		case err != nil && !azure.ResourceNotFound(err):
+			return errors.Wrapf(err, "failed to get nat gateway %s in %s", natGatewaySpec.Name, s.Scope.ResourceGroup())
+		case err == nil:
+			// nat gateway already exists
+			s.Scope.V(4).Info("nat gateway already exists", "nat gateway", natGatewaySpec.Name)
+			s.Scope.SetNodeNatGateway(*existingNatGateway)
```

Comment:
Should this also get called in the case of the NAT Gateway does not exist and needs to be created?

Reply:
I don't know TBH. I saw that it's only done when it already exists for other resources. What should we do?

Reply:
I believe what will happen is that the initial reconcile of `AzureCluster` will create the Nat Gateway, but when the reconciliation is finished, the `AzureCluster` resource will not have the Nat Gateway populated on the `AzureCluster` resource and will be saved in: https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/27cd557e53306d15857d975c8835f9b641a13771/azure/scope/cluster.go#L531-L534

On the second reconciliation, the Nat Gateway will be fetched from Azure and populated via `SetNodeNatGateway`, then it will be persisted to the `AzureCluster` spec when `Close()` is called again.

I've been using a defer to do this elsewhere to ensure we set scope items that are later persisted in `scope.Close()`. For example, scale sets:
https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/27cd557e53306d15857d975c8835f9b641a13771/azure/services/scalesets/scalesets.go#L92-L105

Reply:
sgtm